### PR TITLE
[MatMul] Skew matmul main loop for optimal instruction schedule

### PR DESCRIPTION
**Pull request by shmsong**
**Some recap to understand this PR:**

"Optimized“ matmul kernel in compute bound regime is double buffered in two places:

1. global to shared data transfer (prolog) is double buffered in shared memory. (double buffer1)
2. load from shared memory to register is double buffered in register to interleave mma instructions with shared memory load. (double buffer 2)

In pseudo code the kernel looks like:

```cpp
// double buffer1 prolog
for main_loop_k in ...
  load_CTA_into_smem

// double buffer1 main loop
for main_loop_k in 0...CTA_K:   //loop 1
  load_next_CTA_tile_into_smem

  // double buffer2 prolog
  for j in ...
    load_operand_into_reg
  
  // doublebuffer 2 main loop
  for j in 1...Warp_K:
    load_next_operand_into_reg
    mma

  // double buffer2 epilog
  mma

  __syncthreads()
```

**What this PR does:**

The underlying optimizations in nvcc is really great at optimizing loops that starts with an mma op and finishes with an mma op. So this PR just transforms the kernel code structure above as below:

```cpp
// double buffer1 prolog
for main_loop_k in ...
  load_CTA_into_smem

// vvvvvvvvv added 
// double buffer2 prolog 
//  for iteration 0 of main_loop_k
for j in ...
  load_operand_into_reg
// ^^^^^^^^^^^^^^^^^^^^^^^^^

// double buffer1 main loop
for main_loop_k in 0...CTA_K: 
  load_next_CTA_tile_into_smem

  // doublebuffer 2 main loop
  for j in 1...Warp_K:
    load_next_operand_into_reg
    mma

  // double buffer2 epilog
  mma

  __syncthreads()

   // vvvvvvvvv added 
   // double buffer2 prolog for next main_loop_k iteration
  for j in ...
    load_operand_into_reg_next_iter
  // ^^^^^^^^^^^^^^^^^^^^^^^^^
```
i.e. the doublebuffer2 prolog is skewed by 1 main_loop_k  iteration ahead so that double buffer2's main loop and the mma ops within are able to start at the beginning of the main loop, which gets good optimizations down the stack.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu [hunk 1]
```diff
@@ -56,10 +56,12 @@ __device__ double uniform(unsigned int x, unsigned int y) {
...
-__device__ double randLike(const uint4 &rng_result, int rng_component) {
-  return uniform((&rng_result.x)[rng_component * 2], (&rng_result.x)[rng_component * 2 + 1]);
+__device__ double randLike(const uint4& rng_result, int rng_component) {
+  return uniform(
+      (&rng_result.x)[rng_component * 2],
+      (&rng_result.x)[rng_component * 2 + 1]);
 }
 
-__device__ float randLikef(const uint4 &rng_result, int rng_component) {
+__device__ float randLikef(const uint4& rng_result, int rng_component) {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/runtime/random_numbers.cu]: Just clang-format?

File: torch/csrc/jit/codegen/cuda/type.h [hunk 2]
```diff
@@ -325,7 +325,18 @@ enum class LoadStoreOpType { LdMatrix, LdMatrixTranspose, CpAsync };
...
 // Used to label what part of the double buffered iterdomain
 //  a for loop is materializing.
-enum class DoubleBufferLoopStage { NotApplicable, Prolog, Main, Epilog };
+enum class DoubleBufferLoopStage {
+  NotApplicable,
+  Prolog,
+  Main,
+  Epilog,
+  UpperProlog,
+  LowerProlog
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/type.h:332]: Can you please add comments on what these new variants mean?

**Comment by naoyam** [comment]
Can you please add a few tests on the skewed double buffering transformation?

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp [hunk 3]
```diff
@@ -1158,6 +1159,34 @@ void TensorView::circularBuffer(unsigned int stage) {
   circular_buffer_stage_ = stage;
 }
 
+namespace {
+
+void validateSkewDoubleBuffer(TensorView* tv) {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/tensor_view.cpp:1383]: It wasn't clear until this point what the requirements of using `skewDoubleBuffer` are. 

It seems that those requirements are:

- The tv itself is already set to be double buffered
- The producer of the tv is also double buffered.

I guess we may also want to validate more, for example, making sure the producer tensor is not skewed, as I assume this skewed double buffering won't work in that case.

**Review by naoyam — COMMENTED** [review]
Just a few comments for now. We should definitely have unit tests.

File: torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp [hunk 4]
```diff
@@ -568,6 +912,74 @@ void DoubleBufferInfo::build(Fusion* fusion) {
...
+    bool double_buffer_axis_found = false;
+    for (auto id_it = tv->domain()->domain().rbegin();
+         id_it != tv->domain()->domain().rend();
+         id_it++) {
+      // The outer loop to lift prolog out of would
+      //  be the first serial loop on the left of
+      //  the double buffer loop
+      if (double_buffer_axis_found &&
+          (*id_it)->getParallelType() == ParallelType::Serial) {
+        outer_loop_id = *id_it;
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp:937]: Don't we need to check if that domain is the double buffer loop of the producer?

File: torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp [hunk 5]
```diff
@@ -552,6 +552,350 @@ class DoubleBufferInserter : private kir::ExprMutator {
...
+    auto cloned_loop = IrBuilder::create<kir::ForLoop>(loop);
+    for (auto expr : loop->body().exprs()) {
+      cloned_loop->body().push_back(cloneMaybeLoopNest(expr));
+    }
+    return cloned_loop;
+  }
+
+  // Makes the given vector of for loops into a loop nest,
+  // Returns <Outermost, Innermost> level as a pair.
+  std::pair<kir::ForLoop*, kir::ForLoop*> makeLoopNest(
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp:857]: There's a similar util function

File: torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp [hunk 6]
```diff
@@ -552,6 +552,350 @@ class DoubleBufferInserter : private kir::ExprMutator {
...
+        getClonedPrologLoopNest(original_prolog, upper_prolog_loop);
+
+    // Put the cloned loop nest into the cloned main loop
+    //  and insert the main loop.
+    cloned_main_loop->body().push_back(outer_loop);
+
+    return cloned_main_loop;
+  }
+
+  // Makes the upper prolog loop nest to be inserted at the end of
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp:780]: Typo: lower

File: torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp [hunk 7]
```diff
@@ -552,6 +552,350 @@ class DoubleBufferInserter : private kir::ExprMutator {
...
+//    // consumer could not have been circular buffered
+//    //  as it's a consumer so it's not a cp.async output,
+//    //  which is the only case we have so far (sm80) that
+//    //  can benefit from circular buffering.
+//
+//    for j in 0..1: // Id1 prolog
+//      consumer[j] = producer[i, j]
+//
+//    for j in 0..Id1.size-1: //Id1 main
+//      consumer[j] = producer[i, j]
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp:586]: consumer[j+1] = producer[i, j+1]?

File: torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp [hunk 8]
```diff
@@ -552,6 +552,350 @@ class DoubleBufferInserter : private kir::ExprMutator {
...
+//    for j in 0..Id1.size:
+//      producer [i+1 % stage_depth,...] = ...;
+//
+//    // consumer could not have been circular buffered
+//    //  as it's a consumer so it's not a cp.async output,
+//    //  which is the only case we have so far (sm80) that
+//    //  can benefit from circular buffering.
+//
+//    for j in 0..Id1.size-1: //Id1 main
+//      consumer[j] = producer[i, j]
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp:611]: Same here

File: torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp [hunk 9]
```diff
@@ -552,6 +552,350 @@ class DoubleBufferInserter : private kir::ExprMutator {
...
+        GpuLower::current()->kernel()->oneVal(),
+        false,
+        nullptr,
+        outer_main_loop_->isUnrollRequired(),
+        kir::LoopTransformInfo());
+
+    // Make the upper prolog loop object.
+    auto upper_prolog_loop = IrBuilder::create<kir::ForLoop>(
+        original_prolog->iter_domain(),
+        original_prolog->index(),
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp:759]: Shouldn't this be zero as well?

File: torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp [hunk 10]
```diff
@@ -552,6 +552,350 @@ class DoubleBufferInserter : private kir::ExprMutator {
...
+  }
+
+  // Makes the upper prolog loop nest to be inserted at the end of
+  //  the outer_main_loop (Id0 loop) body.
+  kir::ForLoop* makeLowerProlog(kir::ForLoop* original_prolog) {
+    auto lower_prolog_loop = IrBuilder::create<kir::ForLoop>(
+        original_prolog->iter_domain(),
+        original_prolog->index(),
+        original_prolog->start(),
+        original_prolog->stop(),
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp:787]: According to the above comments, shouldn't this be one?

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp [hunk 11]
```diff
@@ -216,6 +216,9 @@ TensorView::TensorView(const TensorView* src, IrCloner* ir_cloner)
       circular_buffer_stage_(src->circular_buffer_stage_),
       cpu_scalar_(src->cpu_scalar_),
       has_swizzle_op_(src->has_swizzle_op_),
+      lift_read_address_(src->lift_read_address_),
+      lift_write_address_(src->lift_write_address_),
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/tensor_view.cpp:220]: Aha, I think this should go to https://github.com/csarofeen/pytorch/pull/1971

**Pull request closed without merging**